### PR TITLE
Add Input Restriction for safer inputs

### PR DESCRIPTION
**Pull request by X9VoiD**
## Changes

- Made all user-editable text boxes (except Configuration Editor) restricted to numbers only where suitable
- Made OutputModeSelector readonly
- Added `InputRestrictionAttribute` for plugins, which support
    - `RestrictionType.Number`
    - `RestrictionType.Hex`
    - `RestrictionType.Custom`

## Fixes

- UX crash when typing on OutputModeSelector
- UX crash / unstable state when applying or saving invalid input
- Device string reader STAThread access penalty on Windows platforms

### CODE REVIEW

File: OpenTabletDriver.Plugin/Attributes/InputRestrictionAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    public class InputRestrictionAttribute : ModifierAttribute
+    {
+        public InputRestrictionAttribute(InputRestriction restriction, Restrictor customRestrictor)
+        {
+            Restriction = restriction;
+            CustomRestrictor = customRestrictor;
+        }
+
+        public InputRestriction Restriction { private set; get; }
+        public Restrictor CustomRestrictor { private set; get; }
```
Comment (InfinityGhost) [inline OpenTabletDriver.Plugin/Attributes/InputRestrictionAttribute.cs:23]: Could be replaced with `Func<bool, string>`

File: OpenTabletDriver.Plugin/Attributes/InputRestrictionAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+using System;
+
+namespace OpenTabletDriver.Plugin.Attributes
+{
+    public enum InputRestriction
+    {
+        Number,
+        Custom
+    }
```
Comment (InfinityGhost) [inline OpenTabletDriver.Plugin/Attributes/InputRestrictionAttribute.cs:9]: Move this to its own file

File: OpenTabletDriver.UX/Controls/PluginSettingsEditor.cs [hunk 3]
```diff
@@ -240,7 +235,26 @@ private Control ApplyModifierAttributes(Control control, ModifierAttribute attri
...
+                                textBox.TextChanging += (_, args) =>
+                                {
+                                    args.Cancel = inputAttr.CustomRestrictor(args.NewText);
+                                };
+                                break;
+                        }
+                        return textBox;
+                    }
+                    return control;
+
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/PluginSettingsEditor.cs:257]: Wrap this in parenthesis like this, along with the other cases. It allows us to reuse variable names and IMO looks prettier lol
```cs
case InputRestrictionAttribute inputAttr:
{
    if (control is TextBox textBox)
    {
        switch (inputAttr.Restriction)
        {
            case InputRestriction.Number:
                textBox.TextChanging += UXTools.RestrictToNumber;
                break;
            case InputRestriction.Custom:
                textBox.TextChanging += (_, args) =>
                {
                    args.Cancel = inputAttr.CustomRestrictor(args.NewText);
                };
                break;
        }
        return textBox;
    }
    return control;
}
```

File: OpenTabletDriver.UX/Tools/UXTools.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+using System.Text.RegularExpressions;
+
+namespace OpenTabletDriver.UX.Tools
+{
+    public static class UXTools
+    {
+        public static void RestrictToNumber(object _, TextChangingEventArgs args)
+        {
+            args.Cancel = !Regex.IsMatch(args.NewText, "^-*[0-9]*[\\.,]*[0-9]*$");
+        }
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Tools/UXTools.cs:11]: Move this to the `NumberBox` class, I don't see this being used outside of that.

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

**Comment by X9VoiD** [comment]
Changes pushed

File: OpenTabletDriver.Plugin/Attributes/InputRestrictionAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+namespace OpenTabletDriver.Plugin.Attributes
+{
+    [AttributeUsage(AttributeTargets.Property)]
+    public class InputRestrictionAttribute : ModifierAttribute
+    {
+        public InputRestrictionAttribute(InputRestriction restriction, Func<string, bool> customRestrictor)
+        {
+            Restriction = restriction;
+            CustomRestrictor = customRestrictor;
+        }
```
Comment (InfinityGhost) [inline OpenTabletDriver.Plugin/Attributes/InputRestrictionAttribute.cs:12]: Might be a good idea to add another constructor that leaves `customRestrictor` null

File: OpenTabletDriver.UX/Tools/NumberBox.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        {
+            TextChanging += RestrictToNumber;
+        }
+
+        public static void RestrictToNumber(object _, TextChangingEventArgs args)
+        {
+            args.Cancel = !Regex.IsMatch(args.NewText, "^-*[0-9]*[\\.,]*[0-9]*$");
+        }
+    }
+}
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Tools/NumberBox.cs:18]: This belongs in the `OpenTabletDriver.UX.Controls` namespace

**Comment by X9VoiD** [comment]
Changes pushed and tested. `RestrictionType.Hex` for plugins also tested.

**Comment by X9VoiD** [comment]
PR ready for merge

File: OpenTabletDriver.UX/Controls/HexBox.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            get
+            {
+                try
+                {
+                    return Convert.ToInt32(this.Text);
+                }
+                catch
+                {
+                    return 0;
+                }
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/HexBox.cs:20]:
```cs
int.TryParse(this.Text, out var val) ? val : 0;
```

File: OpenTabletDriver.UX/Controls/NumberBox.cs [hunk 8]
```diff
@@ -1,18 +1,12 @@
 using System.Text.RegularExpressions;
-using Eto.Forms;
+using OpenTabletDriver.UX.Controls;
 
-namespace OpenTabletDriver.UX.Controls
+namespace OpenTabletDriver.UX.Tools
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/NumberBox.cs:4]: Why did you put this in tools lol

File: OpenTabletDriver.UX/Controls/HexBox.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+        public override bool Restrictor(string str) => StaticRestrictor(str);
+        public static bool StaticRestrictor(string str)
+        {
+            if (str == "0" || str == "0x")
+            {
+                str = "0x0";
+            }
+
+            return !Regex.IsMatch(str, "^0x[0-9a-f]+$");
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/HexBox.cs:32]: Regex is slower than directly trying to parse
```cs
int.TryParse(str.Replace("0x", string.Empty), NumberStyles.HexNumber, null, out var value) ? value : 0;
```

File: OpenTabletDriver.UX/Controls/HexBox.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
+using System;
+using System.Text.RegularExpressions;
+using OpenTabletDriver.UX.Controls;
+
+namespace OpenTabletDriver.UX.Tools
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/HexBox.cs:5]: Why is this in the tools namespace?

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

File: OpenTabletDriver.UX/Controls/HexBox.cs (hunk 9 above)
Comment (InfinityGhost) [reply OpenTabletDriver.UX/Controls/HexBox.cs]: lg2m

Comment (InfinityGhost) [reply OpenTabletDriver.UX/Controls/HexBox.cs]: Just use `string.Empty` instead of `""`

File: OpenTabletDriver.Daemon/DriverDaemon.cs [hunk 11]
```diff
@@ -277,12 +277,16 @@ private void SetToolSettings()
                     if (property.GetCustomAttribute<PropertyAttribute>(false) != null && 
                         Settings.PluginSettings.TryGetValue(type.FullName + "." + property.Name, out var strValue))
                     {
+                        dynamic value;
```
Comment (InfinityGhost) [inline OpenTabletDriver.Daemon/DriverDaemon.cs:280]: Why not just define this as an object?

Comment (X9VoiD) [reply OpenTabletDriver.Daemon/DriverDaemon.cs]: I did not know at the time that an object will suffice. Change pushed.

File: OpenTabletDriver.UX/Controls/HexBox.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public override int Value => int.TryParse(Regex.Replace(Text, "^0x", string.Empty), NumberStyles.HexNumber, null, out var value) ? value : 0;
+        public override bool Restrictor(string str) => StaticRestrictor(str);
+
+        public static bool StaticRestrictor(string str)
+        {
+            if (str == string.Empty)
+                return false;
+            return !int.TryParse(Regex.Replace(str, "^0x", string.Empty), NumberStyles.HexNumber, null, out var _);
+        }
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/HexBox.cs:16]: Improve naming to something more comprehensive, make the regex static.

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

**Comment by InfinityGhost** [comment]
Should we hold this off until v0.4.3?

**Comment by X9VoiD** [comment]
As discussed, this PR will be closed and branch will be kept for reference. Proposed changes will be rewritten from scratch.

**Pull request closed without merging**